subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java

```diff
@@ -134,10 +191,17 @@ default void visitIdentityInputs(InputVisitor visitor) {}
      * Visit regular inputs of the work.
      *
      * Regular inputs are inputs that are not used to calculate the identity of the work, but used to check up-to-dateness or to calculate the cache key.
-     * To visit all inputs one must call both {@link #visitIdentityInputs(InputVisitor)} and {@link #visitRegularInputs(InputVisitor)}.
+     * To visit all inputs one must call both {@link #visitIdentityInputs(InputVisitor)} as well as this method.
      */
     default void visitRegularInputs(InputVisitor visitor) {}
 
+    /**
+     * Visit outputs of the work in the given workspace.
+     * <p>
+     * Note: For tasks {@code workspace} is {@code null} for now.
```

Comment:
Should there be an `@Nullable` annotation on the `workspace` parameter?

Reply:
I'm not sure what would be the best course of action here. Tasks are executed without a workspace for now, so passing in `null` kind of makes sense. There might be another alternative I'm missing.

Another option would be to assume that a task's workspace is its project build directory. I think it'd make sense to pass that in, but I can't think of a way to utilize it.

That said, all the other `UnitOfWork` implementations operate in the context of a workspace. Enforcing nullability (and wrapping the workspace in an `Optional`) would be additional overhead for all of those guys. That's why I didn't smooth this out yet.

This all might change if/when we introduce auto assignment of output locations (and/or sandboxing), but even then there will be legacy tasks for a long time that will not use these features, and thus will rely on no workspace.

Ideas?

Reply:
Instead of passing a file, you could pass a `Supplier<File>` which is always non-null and which throws when you try to get the workspace from it and you are in the context of a task. I leave it up to you if you think that would be better than possibly throwing an NPE here.